File: datacontrollertests/src/test/java/com/fuzz/datacontroller/test/DataControllerTest.java
package com.fuzz.datacontroller.test;

import com.fuzz.datacontroller.DataController;
import com.fuzz.datacontroller.DataControllerResponse;
import com.fuzz.datacontroller.source.DataSource;
import com.fuzz.datacontroller.source.DataSourceContainer.DataSourceParams;
import com.fuzz.datacontroller.source.MemorySource;

import org.junit.Before;
import org.junit.Test;
import org.mockito.ArgumentCaptor;

import java.util.ArrayList;
import java.util.List;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

/**
 * Description: Tests populating a {@link DataController} and asserts it is correct.
 */
public class DataControllerTest {

    private DataController<String> dataController;
    public DataSource<String> mockDataSource;

    @Before
    public void before_test_Init() {
        //noinspection unchecked
        mockDataSource = mock(DataSource.class);
        when(mockDataSource.sourceType()).thenReturn(DataSource.SourceType.NETWORK);

        when(mockDataSource.refreshStrategy()).thenReturn(new DataSource.RefreshStrategy<String>() {
            @Override
            public boolean shouldRefresh(DataSource<String> dataSource) {
                return true;
            }
        });

        dataController = new DataController.Builder<String>()
                .dataSource(mockDataSource)
                .dataSource(MemorySource.<String>builderInstance().build())
                .build();

    }

    @Test
    public void test_Sources() {
        List<DataSource<String>> sources = new ArrayList<>(dataController.dataSources());
        assertEquals(2, sources.size());

        assertEquals(DataSource.SourceType.MEMORY, sources.get(0).sourceType());
        assertEquals(DataSource.SourceType.NETWORK, sources.get(1).sourceType());
    }

    @Test
    public void test_callSpecific() {
        ArgumentCaptor<DataSource.SourceParams> sourceParamsArgumentCaptor
                = ArgumentCaptor.forClass(DataSource.SourceParams.class);
        ArgumentCaptor<DataController.Success<String>> successArgumentCaptor
                = ArgumentCaptor.forClass(DataController.Success.class);
        ArgumentCaptor<DataController.Error> errorArgumentCaptor
                = ArgumentCaptor.forClass(DataController.Error.class);

        DataSourceParams networkParams = DataSourceParams.networkParams();

        dataController.request(networkParams)
                .build()
                .execute();


        verify(mockDataSource).get(sourceParamsArgumentCaptor.capture(), successArgumentCaptor.capture(),
                errorArgumentCaptor.capture());

        dataController.request(DataSourceParams.memoryParams())
                .build()
                .execute();

        ArgumentCaptor<DataControllerResponse<String>> captor
                = ArgumentCaptor.forClass(DataControllerResponse.class);
        verify(mockDataSource).store(captor.capture());

        dataController.getDataSource(networkParams).getStoredData(null);
        verify(mockDataSource).getStoredData(null);
    }

    @Test
    public void test_requestInvalidSource() {
        boolean failed = false;
        try {
            dataController.request(DataSourceParams.diskParams())
                    .sourceParams(null)
                    .build().execute();
        } catch (RuntimeException r) {
            failed = true;
        }
        assertTrue("Expected a failure for requesting invalid sourcetype: "
                + DataSource.SourceType.DISK, failed);
    }
}

File: datacontrollertests/src/test/java/com/fuzz/datacontroller/DataControllerTest.kt
package com.fuzz.datacontroller

import com.fuzz.datacontroller.DataController
import com.fuzz.datacontroller.DataControllerResponse
import com.fuzz.datacontroller.source.DataSource
import com.fuzz.datacontroller.source.DataSourceContainer.DataSourceParams
import com.fuzz.datacontroller.source.MemorySource
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import org.mockito.ArgumentCaptor
import org.mockito.Mockito.mock
import org.mockito.Mockito.verify
import java.util.*

/**
 * Description: Tests populating a [DataController] and asserts it is correct.
 */
@Suppress("UNCHECKED_CAST")
class DataControllerTest {

    private lateinit var dataController: DataController<String>
    private lateinit var mockDataSource: DataSource<String>
    private lateinit var mockCaller: DataSource.DataSourceCaller<String>
    private lateinit var mockStorage: DataSource.DataSourceStorage<String>

    @Before
    fun before_test_Init() {
        mockCaller = mock<DataSource.DataSourceCaller<*>>(DataSource.DataSourceCaller::class.java)
                as DataSource.DataSourceCaller<String>
        mockStorage = mock<DataSource.DataSourceStorage<*>>(DataSource.DataSourceStorage::class.java)
                as DataSource.DataSourceStorage<String>
        mockDataSource = DataSource.Builder(mockCaller, DataSource.SourceType.NETWORK)
                .storage(mockStorage).build()

        dataController = DataController.Builder<String>()
                .dataSource(mockDataSource)
                .dataSource(MemorySource.builderInstance<String>()
                        .build())
                .build()

    }

    @Test
    fun test_Sources() {
        val sources = ArrayList(dataController.dataSources())
        assertEquals(2, sources.size.toLong())

        assertEquals(DataSource.SourceType.MEMORY, sources[0].sourceType())
        assertEquals(DataSource.SourceType.NETWORK, sources[1].sourceType())
    }

    @Test
    fun test_callSpecific() {
        val sourceParamsArgumentCaptor = ArgumentCaptor.forClass(DataSource.SourceParams::class.java)
        val successArgumentCaptor = ArgumentCaptor.forClass(DataController.Success::class.java)
                as ArgumentCaptor<DataController.Success<String>>
        val errorArgumentCaptor = ArgumentCaptor.forClass(DataController.Error::class.java)

        val networkParams = DataSourceParams.networkParams()

        dataController.request(networkParams).build().execute()


        verify<DataSource.DataSourceCaller<String>>(mockCaller)
                .get(sourceParamsArgumentCaptor.capture(), errorArgumentCaptor.capture(),
                        successArgumentCaptor.capture())

        dataController.request(DataSourceParams.memoryParams())
                .build()
                .execute()

        val captor = ArgumentCaptor.forClass(DataControllerResponse::class.java)
                as ArgumentCaptor<DataControllerResponse<String>>
        verify(mockStorage).store(captor.capture())

        dataController.getDataSource(networkParams).getStoredData(null)
        verify(mockStorage).getStoredData(null)
    }

    @Test
    fun test_requestInvalidSource() {
        var failed = false
        try {
            dataController.request(DataSourceParams.diskParams())
                    .sourceParams(null).build()
                    .execute()
        } catch (r: RuntimeException) {
            failed = true
        }

        assertTrue("Expected a failure for requesting invalid sourcetype: " + DataSource.SourceType.DISK, failed)
    }
}
